### PR TITLE
Add ability to modify ShadingRatePattern as bitmap

### PR DESCRIPTION
**Pull request by bjoeris**
Add ability to modify ShadingRatePattern as bitmap

### CODE REVIEW

**Review by apazylbe — APPROVED** [review]

**Review by Keenuts — APPROVED** [review]
LGTM for the direction, just few comments.

File: src/ppx/grfx/vk/vk_shading_rate.cpp [hunk 1]
```diff
@@ -22,6 +22,117 @@
...
+                // This shading rate is not supported. Find the largest
+                // supported shading rate where neither width nor height is
+                // greater than this fragment size.
+                //
+                // Ties are broken lexicographically, e.g. if 2x2, 1x4 and 4x1
+                // are supported, then 2x4 will be mapped to 2x2 but 4x2 will
+                // map to 4x1.
+                if (width == 1) {
+                    // Width is minimum, can only shrink height.
+                    mMapRateToSupported[encoded] = mMapRateToSupported[RawEncode(width, height - 1)];
```
[the remark this answers is not in the document]
Comment (bjoeris) [reply src/ppx/grfx/vk/vk_shading_rate.cpp:78]: The spec says 1x1 fragment size must always be supported: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetPhysicalDeviceFragmentShadingRatesKHR.html.

But I agree that an assert here makes sense, added.

File: src/ppx/grfx/vk/vk_shading_rate.cpp [hunk 2]
```diff
@@ -22,6 +22,117 @@
...
+}
+
+uint32_t VRSShadingRateEncoder::RawEncode(uint8_t width, uint8_t height)
+{
+    uint32_t widthEnc  = std::min<uint8_t>(width, 4) / 2;
+    uint32_t heightEnc = std::min<uint8_t>(height, 4) / 2;
+    return (widthEnc << 2) | heightEnc;
+}
+
+void VRSShadingRateEncoder::Initialize(const ShadingRateCapabilities& capabilities)
```
[the remark this answers is not in the document]
Comment (bjoeris) [reply src/ppx/grfx/vk/vk_shading_rate.cpp:49]: Yes, I added unit tests for these.

File: src/ppx/grfx/vk/vk_shading_rate.cpp [hunk 3]
```diff
@@ -22,6 +22,117 @@
 namespace ppx {
 namespace grfx {
 namespace vk {
+namespace internal {
+uint32_t FDMShadingRateEncoder::EncodeFragmentDensityImpl(uint8_t xDensity, uint8_t yDensity)
```
[the remark this answers is not in the document]
Comment (bjoeris) [reply src/ppx/grfx/vk/vk_shading_rate.cpp:27]: Added.

File: include/ppx/grfx/vk/vk_shading_rate.h [hunk 4]
```diff
@@ -151,9 +199,11 @@ class ShadingRatePattern
         VkAttachmentReference2                 mVrsAttachmentRef  = {VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_2_KHR};
     };
 
-    Result         CreateApiObjects(const ShadingRatePatternCreateInfo* pCreateInfo) override;
-    void           DestroyApiObjects() override;
-    VkImageViewPtr mAttachmentView;
+    Result CreateApiObjects(const ShadingRatePatternCreateInfo* pCreateInfo) override;
+    void   DestroyApiObjects() override;
+
+    std::shared_ptr<ShadingRateEncoder> mShadingRateEncoder;
```
[the remark this answers is not in the document]
Comment (bjoeris) [reply include/ppx/grfx/vk/vk_shading_rate.h]: Yeah, unique_ptr is probably cleaner. Switched to unique_ptr.

**Pull request merged**